test/units/modules/test_iptables.py

```diff
@@ -947,66 +947,71 @@ def test_destination_ports(self):
-        commands_results = [
-            (0, '', ''),
+        tests = [
+            [
+                {
+                    "chain": "INPUT",
+                    "protocol": "tcp",
+                    "match_set": "admin_hosts",
+                    "match_set_flags": "src",
+                    "destination_port": "22",
+                    "jump": "ACCEPT",
+                    "comment": "this is a comment",
+                },
+                [
+                    "/sbin/iptables", "-t", "filter",
+                    "-C", "INPUT", "-p", "tcp",
+                    "-j", "ACCEPT", "--destination-port", "22",
+                    "-m", "set", "--match-set", "admin_hosts",
+                    "src", "-m", "comment", "--comment", "this is a comment",
+                ],
+            ],
+            [
+                {
+                    "chain": "INPUT",
+                    "protocol": "udp",
+                    "match_set": "banned_hosts",
+                    "match_set_flags": "src,dst",
+                    "jump": "REJECT",
+                },
+                [
+                    "/sbin/iptables", "-t", "filter",
+                    "-C", "INPUT", "-p", "udp",
+                    "-j", "REJECT", "-m", "set",
+                    "--match-set", "banned_hosts", "src,dst",
+                ],
+            ],
+            [
+                {
+                    "chain": "INPUT",
+                    "protocol": "udp",
+                    "match_set": "banned_hosts",
+                    "match_set_flags": "src,src",
+                    "jump": "REJECT",
+                },
+                [
+                    "/sbin/iptables", "-t", "filter",
+                    "-C", "INPUT", "-p", "udp",
+                    "-j", "REJECT", "-m", "set",
+                    "--match-set", "banned_hosts", "src,src",
+                ],
+            ],
         ]
 
-        with patch.object(basic.AnsibleModule, 'run_command') as run_command:
-            run_command.side_effect = commands_results
-            with self.assertRaises(AnsibleExitJson) as result:
-                iptables.main()
-                self.assertTrue(result.exception.args[0]['changed'])
-
-        self.assertEqual(run_command.call_count, 1)
-        self.assertEqual(run_command.call_args_list[0][0][0], [
-            '/sbin/iptables',
-            '-t', 'filter',
-            '-C', 'INPUT',
-            '-p', 'tcp',
-            '-j', 'ACCEPT',
-            '--destination-port', '22',
-            '-m', 'set',
-            '--match-set', 'admin_hosts', 'src',
-            '-m', 'comment',
-            '--comment', 'this is a comment'
-        ])
-
-        set_module_args({
-            'chain': 'INPUT',
-            'protocol': 'udp',
-            'match_set': 'banned_hosts',
-            'match_set_flags': 'src,dst',
-            'jump': 'REJECT',
-        })
-        commands_results = [
-            (0, '', ''),
-        ]
+        for test in tests:
```

Comment:
To have separate tests in the report, use `@pytest.mark.parametrize`. Otherwise, a failure in one iteration will prevent others from even starting.

Reply:
Can we use `pytest.mark.parametrize` in `unittest` as these tests are written in `unittest`?

Reply:
@Akasurde it depends on whether we inherit from `unittest.TestCase`. If yes ­— it limits the use of some of the pytest features, though, there's hacks like autouse fixtures and parameterizing them. If not — all pytest features use for methods in regular classes (though, I do prefer plain functions over methods because I dislike unnecessary scoping/nesting).

That said, `unittest` has a feature to display subtests. You just have to use something like `with self.subTest(id='a param')` in the loop and every case will show up in the test report as a separate entry. (clarify the syntax in the docs, plz!)

Reply:
https://blog.ganssle.io/articles/2020/04/subtests-in-python.html